### PR TITLE
[inductor] Preserve logical indices for arg reductions

### LINKED ISSUE #193661: [inductor] Wrong argmax after fused trailing-dim reduction on non-contiguous inputs (incl. non-unit-stride reduction window) — generic reduction fusion ignores input strides, CPU and CUDA
### 🐛 Describe the bug

**Relationship to #193492**: this is the same *mechanism* (the fused kernel assumes a contiguous interpretation of the input and ignores the real strides) but a **different lowering** — the generic reduction fusion (`mean`/`var`/`sum`/`logsumexp`/`amax` + `argmax`), not the `_adaptive_avg_pool2d` path. The contiguous guard proposed in #193492 is added at the `_adaptive_avg_pool2d` lowering and does not apply to these graphs; the reduction path needs its own fix/regression coverage. I asked in the #193492 thread whether this path should be tracked separately ([[comment](https://github.com/pytorch/pytorch/issues/193492#issuecomment-5299760647)](https://github.com/pytorch/pytorch/issues/193492#issuecomment-5299760647)); opening this issue here so it doesn't get lost in the pool discussion. Happy to close this and fold everything back into #193492 if maintainers prefer a single tracking issue.

`torch.compile` with the Inductor backend returns a wrong `argmax` index when a trailing-dim reduction (`mean`, `var`, `sum`, `logsumexp`, also `amax`) is fused with `argmax` and the reduction input is a **non-contiguous view**. Eager returns the correct index; the compiled index points at a value that is **not** the extremum of the reduction output. The unfused reduction matches eager exactly (sanity `maxdiff ≈ 0`), so the mis-indexing is introduced by the fused kernel, whose generated input indexing assumes a contiguous layout and ignores the real strides. The trigger is deterministic (fixed seed, no randomness).

The previously-reported family instances used non-contiguous leading dims with a **stride-1** reduction window (e.g. `transpose`). The new instance below additionally has a **non-unit-stride reduction window** (`[..., ::2]`, stride 2 on the reduced dim) combined with non-contiguous leading dims. Neither axis triggers on its own: the same `[..., ::2]` slice on a contiguous input passes, and a fully contiguous input passes. As a scope hint for the regression test, `expand`/stride-0 views are **not** affected (all pass).

### To reproduce

```python
import torch

def fn(x):
    return torch.argmax(torch.mean(x, dim=-1))

torch.manual_seed(123)
x = torch.rand([2, 4, 4, 8]).transpose(0, 1).contiguous().transpose(0, 1)[..., ::2]
# shape (2, 4, 4, 4), strides (32, 64, 8, 2)
# non-contiguous leading dims AND a stride-2 reduction window

print("eager   =", fn(x).item())                # 12 (correct)
print("compiled=", torch.compile(fn)(x).item()) # 24 (wrong)
```

Truth check: value at the returned index vs. the true maximum of `mean(x, dim=-1)` (eager index points at the maximum; compiled index does not):

```text
eager    index 12 -> value 0.828343 (== max)
compiled index 24 -> value 0.596202 (!= max)
```

Scope / additional observations (all fixed seed 123, caches disabled, unfused reductions matching eager exactly):

- The same input mismatches for `var`, `sum`, `logsumexp` (dim=-1) and for `mean`, `sum`, `var` (dim=[-1, -2]): **7/7 reduction variants**, on CPU and CUDA, torch 2.13.0 and 2.5.1 (CPU indices identical across versions, e.g. mean: eager 12 / compiled 24 / true 12; CUDA also reproduces).
- Controls: the same `[..., ::2]` slice on a contiguous input passes, a fully contiguous input passes, and `expand`/stride-0 views pass — the trigger is the combination of non-contiguous leading dims and a non-unit-stride reduction window.
- The earlier family instances (non-contiguous leading dims, stride-1 reduction window) are also affected; see the #193492 thread for the family matrix.

Kernel evidence (CPU, torch 2.13.0): the aten graph contains only `mean.dim -> argmax` (the transpose/slice views are metadata and never enter the graph), while the fused kernel `cpp_fused_argmax_mean_0` reads the reduction window with a flat, contiguous interpretation:

```cpp
for (int64_t x0 = 0; x0 < 32; x0++) {   // x0 = flattened 2*4*4 leading positions
    auto tmp0 = in_ptr0[8 * x0];        // mean over in_ptr0[8*x0 + {0,2,4,6}]
    auto tmp1 = in_ptr0[2 + 8 * x0];
    auto tmp3 = in_ptr0[4 + 8 * x0];
    auto tmp5 = in_ptr0[6 + 8 * x0];
    ...
    tmp_acc0 = argmax_combine(tmp_acc0, mean, x0);
}
```

The true offset of element (i, j, k, k′) with strides `(32, 64, 8, 2)` is `32i + 64j + 8k + 2k′`, which the fused flat reads do not match. The wrong reads stay in bounds (max flat offset 254 < storage size 256), so the result is silently wrong — no OOB, no exception.

### Error logs

No exception — silent wrong result (wrong index returned).

### Versions

```text
PyTorch version: 2.13.0+cu130 and 2.5.1+cu121 (both reproduce, CPU and CUDA)
OS: Linux (AutoDL container)
```

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by eellison**
We have an optimization that replaces an argmax `(value, index)` pair with the loop index when iterating contiguously. Previously, that optimization ran at lowering time by inspecting input strides. However, this decision must compose with subsequent optimizations such as layout changes and loop reordering.

Instead, always preserve the logical index through lowering, even when inputs are contiguous, and move the optimization immediately before codegen, after loop transformations are complete.

> AI-assisted implementation details:
>
> The late optimization removes the explicit logical index only when it proves that the index matches the final native reduction order. Otherwise, codegen retains the explicit index. Triton reduction CSE also includes retained logical indices, so equivalent mappings can merge while distinct mappings remain separate.
>
> Fixes #193661. Also covers #193751.
>
> The implementation was prepared with an AI assistant and reviewed and validated by the author.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -16919,12 +16919,18 @@ def f():
...
     def test_adaptive_avg_pool1d_argmax(self):
         # https://github.com/pytorch/pytorch/issues/113013
+        # https://github.com/pytorch/pytorch/issues/193492
         def fn(x):
             x = torch.adaptive_avg_pool1d(input=x, output_size=2)
             x = torch.argmax(input=x)
             return x
 
-        x = torch.rand([4, 4, 3], dtype=torch.float64)
+        x = torch.zeros(4, 4, 3, device=self.device, dtype=torch.float64).transpose(
```
Comment (jananisriram) [inline test/inductor/test_torchinductor.py]: Making this input non-contiguous means CPU-Halide and CPU-Pallas now exercise it, and `supports_logical_index_argreduce` is False on both, so I'd expect 27 rather than 15 there. `@skip_if_gpu_halide` only fires when `device.type == "cuda"`, and I don't see a `pallas_skip_tests/` entry. Would `@skip_if_halide` plus a pallas sentinel cover it?

[from my agent]

File: test/inductor/test_torchinductor.py [hunk 2]
```diff
@@ -18979,6 +18985,57 @@ def f(x):
...
+        self.common(fn, (x,))
+
+        compiled = torch.compile(fn, fullgraph=True)
+        if is_cpp_backend(self.device):
+            _, code = run_and_get_cpp_code(compiled, x)
+            self.assertIn("tmp_acc0", code)
+            self.assertNotIn("tmp_acc1", code)
+        else:
+            code = run_and_get_triton_code(compiled, x)
+            self.assertEqual(code.count("@triton_heuristics."), 1)
```
Comment (jananisriram) [inline test/inductor/test_torchinductor.py]: These counts get copied into the dynamic-shapes suites with `assume_static_by_default=False`. Once `_simplify_loops` rewrites the merged index into ModularIndexing/FloorDiv form, `statically_known_equals` may not prove it equal to `flatten_index`, which would retain the index and give 2 here. Did the dynamic-shapes variants pass? `test_torchinductor_codegen_dynamic_shapes.py:153` already skips the sibling test.

The same mechanism is a perf question: a contiguous multi-dim argmax under dynamic shapes would now materialize an int64 index that main never created.

[from my agent]

File: torch/_inductor/codegen/triton.py [hunk 3]
```diff
@@ -5556,7 +5556,11 @@ def argreduce_result_kind():
                 return "value"
             return "index"
 
-        cache_key = (src_dtype, reduction_type, value)
+        cache_key = (
+            src_dtype,
+            reduction_type,
+            value if logical_index is None else (value, logical_index),
```
Comment (jananisriram) [inline torch/_inductor/codegen/triton.py]: This is the same fix as #194093, which is open and approved — worth linking so one of the two can close. Also worth noting cpp.py:2486 and :3278 already key on the index, so this is bringing Triton to parity rather than introducing a new convention.

[from my agent]

File: torch/_inductor/optimize_indexing.py [hunk 4]
```diff
@@ -10,7 +11,58 @@
...
+                    continue
+                reductions[node] = reduction_value
+                if not V.graph.sizevars.statically_known_equals(
+                    loop_body.indexing_exprs[index_name], native_index
+                ):
+                    non_native.add(node)
+
+    for node, reduction_value in reductions.items():
+        if node not in non_native:
+            node.args = (*node.args[:-1], reduction_value)
```
Comment (jananisriram) [inline torch/_inductor/optimize_indexing.py]: This drops the use but leaves the `index_expr`/`get_index` nodes in the graph, and `LoopBodyBlock.__call__` runs every node through `InterpreterShim` (`garbage_collect_values=False`), so `ops.index_expr` is still dispatched at codegen.

On CPU that isn't free: the vectorization check at cpp.py:4226 walks `sub_block.graph.nodes` directly, so a zero-user `index_expr` whose stride it can't resolve still counts toward `non_contig_indexing_op_counter`, and these bodies have a small `op_num` — one extra op can reach `OVERHEAD_RATIO_THRESHOLD = 0.12` and turn vectorization off for the kernel. I haven't confirmed a case that actually trips it.

Is retention deliberate because `LoopBodyBlock.clone` shares the graph? After phase 1 has cleared every sharing body, "no users" is graph-global, so `erase_node` guarded on `not node.users` looks safe by that same invariant.

[from my agent]

**Comment by jananisriram** [comment]
@eellison might be worth checking whether this PR unifies pytorch#193680, pytorch#194093, and pytorch#195063 (all flagged by my agent as similar).

File: test/inductor/test_torchinductor.py (hunk 1 above)
Comment (jananisriram) [inline test/inductor/test_torchinductor.py]: Making this input non-contiguous means CPU-Halide and CPU-Pallas now exercise it, and `supports_logical_index_argreduce` is False on both, so I'd expect 27 rather than 15 there. `@skip_if_gpu_halide` only fires when `device.type == "cuda"`, and I don't see a `pallas_skip_tests/` entry. Would `@skip_if_halide` plus a pallas sentinel cover it?

[from my agent]

File: torch/_inductor/codegen/triton.py (hunk 3 above)
Comment (jananisriram) [inline torch/_inductor/codegen/triton.py]: This is the same fix as #194093, which is open and approved — worth linking so one of the two can close. Also worth noting cpp.py:2486 and :3278 already key on the index, so this is bringing Triton to parity rather than introducing a new convention.

[from my agent]

File: torch/_inductor/optimize_indexing.py (hunk 4 above)
Comment (jananisriram) [inline torch/_inductor/optimize_indexing.py]: Only the use goes away here — the `index_expr`/`get_index` nodes stay, and `LoopBodyBlock.__call__` interprets every node in the graph, so the index arithmetic is still emitted into the kernel.

On CPU that may cost more than an extra line: the vectorization heuristic (cpp.py:4226) counts those leftover nodes, and an argreduce body is small enough that one more can cross the 0.12 threshold and disable vectorization. I haven't confirmed a case that trips it.

Is keeping them deliberate because `clone` shares the graph between bodies? If so it's worth a comment — otherwise `erase_node` when there are no users looks safe.

[from my agent]

File: test/inductor/test_torchinductor.py (hunk 1 above)
Comment (jananisriram) [inline test/inductor/test_torchinductor.py:16938]: should we skip on all halide (i.e. @skip_if_halide instead of @skip_if_gpu_halide) since non-contiguous input --> CPU-halide exercises the test, but supports_logical_index_argreduce is false?

[from my agent]

File: test/inductor/test_torchinductor.py [hunk 5]
```diff
@@ -18979,6 +18985,57 @@ def f(x):
         self.assertTrue("ReductionHint.OUTER" in code)
         self.assertFalse("ReductionHint.INNER" in code)
 
+    @parametrize("slice_pointwise", (False, True))
+    @skip_if_halide
```
Comment (jananisriram) [inline test/inductor/test_torchinductor.py:18999]: This one gets @skip_if_halide but not @skip_if_pallas, while test_argreduce_native_index_cse just below has both. make_pallas(CpuTests) runs this suite with cpu_backend="pallas", which has no logical-index support either — should this carry the pallas skip as well?

[from my agent]

File: test/inductor/test_torchinductor.py (hunk 2 above)
Comment (jananisriram) [inline test/inductor/test_torchinductor.py:19046]: this could potentially fail with dynamic shapes (assume_static_by_default=False); does the dynamic shape variant run in CI?

**Review by jananisriram — APPROVED** [review]
left a couple minor points but otherwise lgtm, thanks for the fix!

**Review by jansel — APPROVED** [review]

File: test/inductor/test_torchinductor.py (hunk 2 above)
Comment (eellison) [reply test/inductor/test_torchinductor.py:19046]: it does.
